File: src/ipc/listener.ts
// Realtime 리스너 IPC 래퍼 (`docs/03-ipc-contract.md` §8.5 v0.10).

import { call } from "./index";
import { getActiveSessionId } from "@/stores/tabsStore";
import type { ListenerDsl, ListenerInfo } from "@/types";

function requireActiveSessionId(): string {
  const id = getActiveSessionId();
  if (!id) throw new Error("no active session — activate a profile first");
  return id;
}

/**
 * Realtime listener를 등록·시작한다. 결과는 `listener:change:<id>` /
 * `listener:status:<id>` 이벤트로 흐른다.
 */
export const startListener = (listenerId: string, dsl: ListenerDsl) =>
  call<void>("start_listener", {
    session_id: requireActiveSessionId(),
    params: { listener_id: listenerId, dsl },
  });

/** listener를 종료한다. 알 수 없는 id는 idempotent. */
export const stopListener = (listenerId: string) =>
  call<void>("stop_listener", { listener_id: listenerId });

/** 활성 listener 메타데이터 목록 (디버그/세션 종료 시 일괄 처리용). */
export const listListeners = () => call<ListenerInfo[]>("list_listeners", {});

File: src-tauri/src/commands/listener.rs
//! Realtime 리스너 커맨드 (`docs/03-ipc-contract.md` §8.5, Phase 11).
//!
//! 활성 세션의 라이브 `FirestoreDb`를 사용해 `db.create_listener` 기반
//! 리스너를 등록·종료한다. 세션 deactivate 시 자동 정리되므로 여기서는
//! lifecycle을 직접 다루지 않는다.

use std::sync::Arc;

use serde::Deserialize;
use tauri::{AppHandle, State};
use uuid::Uuid;

use crate::error::AppResult;
use crate::firestore::listener::{self, ListenerInfoDto};
use crate::query::dsl::ListenerDsl;
use crate::state::AppState;

#[derive(Deserialize)]
pub struct StartListenerParams {
    pub listener_id: String,
    pub dsl: ListenerDsl,
}

#[tauri::command(rename_all = "snake_case")]
pub async fn start_listener(
    app: AppHandle,
    state: State<'_, AppState>,
    session_id: Uuid,
    params: StartListenerParams,
) -> AppResult<()> {
    let client = state.sessions.firestore(session_id)?;
    let registry = Arc::clone(state.sessions.listeners());
    listener::start_listener(
        app,
        client.db.clone(),
        registry,
        params.listener_id,
        session_id,
        params.dsl,
    )
    .await
}

#[tauri::command(rename_all = "snake_case")]
pub async fn stop_listener(
    state: State<'_, AppState>,
    listener_id: String,
) -> AppResult<()> {
    let registry = Arc::clone(state.sessions.listeners());
    listener::stop_listener(registry, &listener_id).await
}

#[tauri::command(rename_all = "snake_case")]
pub fn list_listeners(state: State<'_, AppState>) -> AppResult<Vec<ListenerInfoDto>> {
    Ok(state.sessions.listeners().list())
}
